File: frontend-offline/src/ai/pathfinder.ts
import { TILE_X_NUM, TILE_Y_NUM } from '../game/gameConfig'
import { MapIndex, PressedDir } from '../game/types'
import { ObjManager } from '../game/objManager'

export interface PathResult {
    dir: PressedDir
    steps: number
}

const DIRS: { dx: number; dy: number; dir: PressedDir }[] = [
    { dx: 0, dy: -1, dir: 'up' },
    { dx: 0, dy: 1, dir: 'down' },
    { dx: -1, dy: 0, dir: 'left' },
    { dx: 1, dy: 0, dir: 'right' },
]

/**
 * BFS from `start` toward `target`. Returns the first direction to take and
 * the total steps, or null if no path exists.
 *
 * When `avoidDanger` is true, tiles marked in dangerMap are skipped.
 */
export function bfsNextDir(
    start: MapIndex,
    target: MapIndex,
    objManager: ObjManager,
    dangerMap: boolean[][],
    avoidDanger: boolean
): PathResult | null {
    if (start.indexX === target.indexX && start.indexY === target.indexY) {
        return { dir: null, steps: 0 }
    }

    const visited = Array.from({ length: TILE_Y_NUM }, () => Array(TILE_X_NUM).fill(false))
    visited[start.indexY][start.indexX] = true

    const queue: { index: MapIndex; dir: PressedDir; steps: number }[] = []

    for (const { dx, dy, dir } of DIRS) {
        const nx = start.indexX + dx
        const ny = start.indexY + dy
        if (outOfBounds(nx, ny) || visited[ny][nx]) continue
        const tileType = objManager.mapManager.getMapTileTypeByIndex({ indexX: nx, indexY: ny })
        if (tileType === 'wall' || tileType === 'brick' || tileType === 'bomb') continue
        if (avoidDanger && dangerMap[ny][nx]) continue
        visited[ny][nx] = true
        queue.push({ index: { indexX: nx, indexY: ny }, dir, steps: 1 })
    }

    while (queue.length > 0) {
        const { index, dir, steps } = queue.shift()!
        if (index.indexX === target.indexX && index.indexY === target.indexY) {
            return { dir, steps }
        }

        for (const { dx, dy } of DIRS) {
            const nx = index.indexX + dx
            const ny = index.indexY + dy
            if (outOfBounds(nx, ny) || visited[ny][nx]) continue
            const tileType = objManager.mapManager.getMapTileTypeByIndex({ indexX: nx, indexY: ny })
            if (tileType === 'wall' || tileType === 'brick' || tileType === 'bomb') continue
            if (avoidDanger && dangerMap[ny][nx]) continue
            visited[ny][nx] = true
            queue.push({ index: { indexX: nx, indexY: ny }, dir, steps: steps + 1 })
        }
    }

    return null
}

/**
 * BFS to find the nearest tile NOT in dangerMap, starting from `start`.
 * Used after placing a bomb to escape the blast zone.
 * Does not skip bomb tiles so the AI can walk through its own bomb.
 */
export function findNearestSafeTile(
    start: MapIndex,
    objManager: ObjManager,
    dangerMap: boolean[][]
): PathResult | null {
    if (!dangerMap[start.indexY][start.indexX]) {
        return { dir: null, steps: 0 }
    }

    const visited = Array.from({ length: TILE_Y_NUM }, () => Array(TILE_X_NUM).fill(false))
    visited[start.indexY][start.indexX] = true

    const queue: { index: MapIndex; dir: PressedDir; steps: number }[] = []

    for (const { dx, dy, dir } of DIRS) {
        const nx = start.indexX + dx
        const ny = start.indexY + dy
        if (outOfBounds(nx, ny) || visited[ny][nx]) continue
        const tileType = objManager.mapManager.getMapTileTypeByIndex({ indexX: nx, indexY: ny })
        if (tileType === 'wall' || tileType === 'brick') continue
        visited[ny][nx] = true
        queue.push({ index: { indexX: nx, indexY: ny }, dir, steps: 1 })
    }

    while (queue.length > 0) {
        const { index, dir, steps } = queue.shift()!
        if (!dangerMap[index.indexY][index.indexX]) {
            return { dir, steps }
        }

        for (const { dx, dy } of DIRS) {
            const nx = index.indexX + dx
            const ny = index.indexY + dy
            if (outOfBounds(nx, ny) || visited[ny][nx]) continue
            const tileType = objManager.mapManager.getMapTileTypeByIndex({ indexX: nx, indexY: ny })
            if (tileType === 'wall' || tileType === 'brick') continue
            visited[ny][nx] = true
            queue.push({ index: { indexX: nx, indexY: ny }, dir, steps: steps + 1 })
        }
    }

    return null
}

function outOfBounds(x: number, y: number) {
    return x < 0 || x >= TILE_X_NUM || y < 0 || y >= TILE_Y_NUM
}

File: py-game/ai/pathfinder.py
from collections import deque
from config import TILE_X_NUM, TILE_Y_NUM
from game.obj_manager import ObjManager

DIRS = [
    (0, -1, "up"),
    (0,  1, "down"),
    (-1, 0, "left"),
    (1,  0, "right"),
]


def _out_of_bounds(x: int, y: int) -> bool:
    return x < 0 or x >= TILE_X_NUM or y < 0 or y >= TILE_Y_NUM


def bfs_next_dir(
    start: tuple[int, int],
    target: tuple[int, int],
    obj_manager: ObjManager,
    danger_map: list[list[bool]],
    avoid_danger: bool,
) -> dict | None:
    """
    BFS from `start` toward `target`.
    Returns {"dir": str, "steps": int} or None if no path found.
    """
    sx, sy = start
    tx, ty = target

    if sx == tx and sy == ty:
        return {"dir": None, "steps": 0}

    visited = [[False] * TILE_X_NUM for _ in range(TILE_Y_NUM)]
    visited[sy][sx] = True
    queue = deque()

    for dx, dy, d in DIRS:
        nx, ny = sx + dx, sy + dy
        if _out_of_bounds(nx, ny) or visited[ny][nx]:
            continue
        tile_type = obj_manager.map_manager.get_tile_type(nx, ny)
        if tile_type in ("wall", "brick", "bomb"):
            continue
        if avoid_danger and danger_map[ny][nx]:
            continue
        visited[ny][nx] = True
        queue.append({"pos": (nx, ny), "dir": d, "steps": 1})

    while queue:
        node = queue.popleft()
        nx, ny = node["pos"]
        if nx == tx and ny == ty:
            return {"dir": node["dir"], "steps": node["steps"]}

        for dx, dy, _ in DIRS:
            nnx, nny = nx + dx, ny + dy
            if _out_of_bounds(nnx, nny) or visited[nny][nnx]:
                continue
            tile_type = obj_manager.map_manager.get_tile_type(nnx, nny)
            if tile_type in ("wall", "brick", "bomb"):
                continue
            if avoid_danger and danger_map[nny][nnx]:
                continue
            visited[nny][nnx] = True
            queue.append({"pos": (nnx, nny), "dir": node["dir"], "steps": node["steps"] + 1})

    return None


def find_nearest_safe_tile(
    start: tuple[int, int],
    obj_manager: ObjManager,
    danger_map: list[list[bool]],
) -> dict | None:
    """
    BFS to find the nearest tile NOT in the danger map.
    The AI may walk through bomb tiles (its own bombs) during escape.
    Returns {"dir": str, "steps": int} or None.
    """
    sx, sy = start
    if not danger_map[sy][sx]:
        return {"dir": None, "steps": 0}

    visited = [[False] * TILE_X_NUM for _ in range(TILE_Y_NUM)]
    visited[sy][sx] = True
    queue = deque()

    for dx, dy, d in DIRS:
        nx, ny = sx + dx, sy + dy
        if _out_of_bounds(nx, ny) or visited[ny][nx]:
            continue
        tile_type = obj_manager.map_manager.get_tile_type(nx, ny)
        if tile_type in ("wall", "brick"):
            continue
        visited[ny][nx] = True
        queue.append({"pos": (nx, ny), "dir": d, "steps": 1})

    while queue:
        node = queue.popleft()
        nx, ny = node["pos"]
        if not danger_map[ny][nx]:
            return {"dir": node["dir"], "steps": node["steps"]}

        for dx, dy, _ in DIRS:
            nnx, nny = nx + dx, ny + dy
            if _out_of_bounds(nnx, nny) or visited[nny][nnx]:
                continue
            tile_type = obj_manager.map_manager.get_tile_type(nnx, nny)
            if tile_type in ("wall", "brick"):
                continue
            visited[nny][nnx] = True
            queue.append({"pos": (nnx, nny), "dir": node["dir"], "steps": node["steps"] + 1})

    return None
